{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Jacob Duvall"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# the_summarizer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import random\n",
    "import glob\n",
    "from sklearn.metrics import silhouette_score\n",
    "from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer\n",
    "from sklearn.cluster import KMeans\n",
    "import json\n",
    "import pickle\n",
    "import nltk\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import networkx\n",
    "import os\n",
    "\n",
    "\n",
    "# look at the format of the json schema\n",
    "def look_at_the_format_of_the_file(file):\n",
    "    file = open(file, 'r')\n",
    "    print(file.read())\n",
    "\n",
    "\n",
    "# Takes glob and randomly extracts a percent of the files in a list\n",
    "def choose_documents(folders, percent):\n",
    "    file_list = list()\n",
    "    for folder in folders:\n",
    "        files = os.listdir(folder)\n",
    "        for file in files:\n",
    "            file_list.append(file)\n",
    "    size = int(len(file_list) * (percent*(10**-2)))\n",
    "    return random.sample(file_list, size)\n",
    "\n",
    "\n",
    "# Takes a list of files and tokenizes the data in the files using CountVectorizer\n",
    "def files_reader(files_list):\n",
    "    directory = 'C:\\\\Users\\\\jdale\\\\OneDrive\\\\School\\\\Text Analytics\\\\' \\\n",
    "                '*\\\\*\\\\pdf_json\\\\'\n",
    "    single_large_document_list = list()\n",
    "    for file in files_list:\n",
    "        dir = glob.glob(str(directory + file))\n",
    "        with open(dir[0]) as file_json:\n",
    "            data = json.load(file_json)\n",
    "            document_string = ''\n",
    "            for line in data['body_text']:\n",
    "                document_string = document_string + line['text']\n",
    "            single_large_document_list.append(document_string)\n",
    "\n",
    "    return single_large_document_list\n",
    "\n",
    "\n",
    "# use tfidfvectorizer on the documents to get the matrix\n",
    "def tfid_vectorize(doc):\n",
    "    tv = TfidfVectorizer(min_df = 0., max_df=1., use_idf=True)\n",
    "    tv_matrix = tv.fit_transform(doc)\n",
    "    pickle.dump(tv_matrix, open('yummy_pickle.pkl', 'wb'))\n",
    "    return tv, tv_matrix\n",
    "\n",
    "\n",
    "# use countvectoizer on the documents to get the matrix\n",
    "def count_vectorize(doc):\n",
    "    cv = CountVectorizer()\n",
    "    cv_matrix = cv.fit_transform(doc)\n",
    "    pickle.dump(cv_matrix, open('yummy_pickle_cv.pkl', 'wb'))\n",
    "    return cv, cv_matrix\n",
    "\n",
    "\n",
    "# Takes tokenized documents and clusters them -- Uses Silhouette Coefficient to measure cluster quality.\n",
    "# Records the documents that are part of each cluster\n",
    "def cluster_documents(doc_list):\n",
    "    cv, tokenized_files = count_vectorize(doc_list)\n",
    "    cluster_range = list(range(2, 8))\n",
    "    for n_clusters in cluster_range:\n",
    "        km = KMeans(n_clusters=n_clusters)\n",
    "        km_predicts = km.fit_predict(tokenized_files)\n",
    "        pickle_save = 'pickle_km_' + str(n_clusters) + '.pkl'\n",
    "        pickle.dump(km_predicts, open(pickle_save, 'wb'))\n",
    "\n",
    "        score = silhouette_score(tokenized_files, km_predicts)\n",
    "        print(\"Number of clusters: {}, Silhouette Score: {}\".format(n_clusters, score))\n",
    "    return cv, tokenized_files\n",
    "\n",
    "\n",
    "# fits the best cluster size and shows the feature names \n",
    "def best_cluster(cv, tokenized_files, size):\n",
    "    km = KMeans(n_clusters=size).fit(tokenized_files)\n",
    "    feature_names = cv.get_feature_names()\n",
    "    ordered_centroids = km.cluster_centers_.argsort()[:, ::-1]\n",
    "    for cluster_num in range(2):\n",
    "        print('CLUSTER #' +str(cluster_num+1))\n",
    "        feature_list = list()\n",
    "        for i in ordered_centroids[cluster_num, :10]:\n",
    "            feature_list.append(feature_names[i])\n",
    "        print(feature_list)\n",
    "\n",
    "\n",
    "# summarize the documents looking at their top 8 sentences through TextRank\n",
    "def summarize_documents(string_list):\n",
    "    count = 0\n",
    "    for string in string_list:\n",
    "        sentences = nltk.sent_tokenize(string)\n",
    "        try:\n",
    "            if len(sentences) < 8:\n",
    "                raise Exception\n",
    "        except Exception:\n",
    "            continue\n",
    "        tv, tv_matrix = tfid_vectorize(sentences)\n",
    "        try:\n",
    "            tv_matrix = tv_matrix.toarray()\n",
    "        except:\n",
    "            print('too large')\n",
    "            continue\n",
    "        similarity_matrix = np.matmul(tv_matrix, tv_matrix.T)\n",
    "        try:\n",
    "            similarity_graph = networkx.from_numpy_array(similarity_matrix)\n",
    "        except MemoryError:\n",
    "            print('memory error 1')\n",
    "            continue\n",
    "        try:\n",
    "            scores = networkx.pagerank(similarity_graph)\n",
    "        except MemoryError:\n",
    "            print('memory error')\n",
    "            continue\n",
    "\n",
    "        ranked_sentences = sorted(((score, index) for index, score in scores.items()), reverse=True)\n",
    "        top_sentenc_indices = [ranked_sentences[index][1] for index in range(8)]\n",
    "\n",
    "        top_sentenc_indices.sort()\n",
    "        write_summary_to_file(count, np.array(sentences)[top_sentenc_indices])\n",
    "        count += 1\n",
    "\n",
    "\n",
    "# write the summaries generated from summarize_documents() to file\n",
    "def write_summary_to_file(count, summary_array):\n",
    "    #print(count)\n",
    "    if count < 1:\n",
    "        try:\n",
    "            os.remove(\"SUMMARY.md\")\n",
    "        except:\n",
    "            pass\n",
    "        file = open(\"SUMMARY.md\", \"w\",  encoding='utf-8')\n",
    "        header = 'This file was generated using TextRank summarization. The process taken was to extract ' \\\n",
    "                 'a random percent sampling of pdf_json files from ' \\\n",
    "                 'https://www.kaggle.com/allen-institute-for-ai/CORD-19-research-challenge. With these ' \\\n",
    "                 'files extracted, I then opened all files, tokenized the sentences of the ' \\\n",
    "                 'files using tfid vectorizer, ' \\\n",
    "                 'and then applied a TextRank algorithm to the tokenized docs. This TextRank allowed me to then ' \\\n",
    "                 'extract the 8 most useful summarized sentences. And voila!'\n",
    "        file.write(header)\n",
    "        file.write('\\n\\n')\n",
    "        file.close()\n",
    "\n",
    "    file = open(\"SUMMARY.md\", \"a\",  encoding='utf-8')\n",
    "    file.write(str(summary_array))\n",
    "    file.write('\\n\\n')\n",
    "    file.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# main"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Look at the format of the file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# JSON schema of full text documents\n",
      "\n",
      "\n",
      "{\n",
      "    \"paper_id\": <str>,                      # 40-character sha1 of the PDF\n",
      "    \"metadata\": {\n",
      "        \"title\": <str>,\n",
      "        \"authors\": [                        # list of author dicts, in order\n",
      "            {\n",
      "                \"first\": <str>,\n",
      "                \"middle\": <list of str>,\n",
      "                \"last\": <str>,\n",
      "                \"suffix\": <str>,\n",
      "                \"affiliation\": <dict>,\n",
      "                \"email\": <str>\n",
      "            },\n",
      "            ...\n",
      "        ],\n",
      "        \"abstract\": [                       # list of paragraphs in the abstract\n",
      "            {\n",
      "                \"text\": <str>,\n",
      "                \"cite_spans\": [             # list of character indices of inline citations\n",
      "                                            # e.g. citation \"[7]\" occurs at positions 151-154 in \"text\"\n",
      "                                            #      linked to bibliography entry BIBREF3\n",
      "                    {\n",
      "                        \"start\": 151,\n",
      "                        \"end\": 154,\n",
      "                        \"text\": \"[7]\",\n",
      "                        \"ref_id\": \"BIBREF3\"\n",
      "                    },\n",
      "                    ...\n",
      "                ],\n",
      "                \"ref_spans\": <list of dicts similar to cite_spans>,     # e.g. inline reference to \"Table 1\"\n",
      "                \"section\": \"Abstract\"\n",
      "            },\n",
      "            ...\n",
      "        ],\n",
      "        \"body_text\": [                      # list of paragraphs in full body\n",
      "                                            # paragraph dicts look the same as above\n",
      "            {\n",
      "                \"text\": <str>,\n",
      "                \"cite_spans\": [],\n",
      "                \"ref_spans\": [],\n",
      "                \"eq_spans\": [],\n",
      "                \"section\": \"Introduction\"\n",
      "            },\n",
      "            ...\n",
      "            {\n",
      "                ...,\n",
      "                \"section\": \"Conclusion\"\n",
      "            }\n",
      "        ],\n",
      "        \"bib_entries\": {\n",
      "            \"BIBREF0\": {\n",
      "                \"ref_id\": <str>,\n",
      "                \"title\": <str>,\n",
      "                \"authors\": <list of dict>       # same structure as earlier,\n",
      "                                                # but without `affiliation` or `email`\n",
      "                \"year\": <int>,\n",
      "                \"venue\": <str>,\n",
      "                \"volume\": <str>,\n",
      "                \"issn\": <str>,\n",
      "                \"pages\": <str>,\n",
      "                \"other_ids\": {\n",
      "                    \"DOI\": [\n",
      "                        <str>\n",
      "                    ]\n",
      "                }\n",
      "            },\n",
      "            \"BIBREF1\": {},\n",
      "            ...\n",
      "            \"BIBREF25\": {}\n",
      "        },\n",
      "        \"ref_entries\":\n",
      "            \"FIGREF0\": {\n",
      "                \"text\": <str>,                  # figure caption text\n",
      "                \"type\": \"figure\"\n",
      "            },\n",
      "            ...\n",
      "            \"TABREF13\": {\n",
      "                \"text\": <str>,                  # table caption text\n",
      "                \"type\": \"table\"\n",
      "            }\n",
      "        },\n",
      "        \"back_matter\": <list of dict>           # same structure as body_text\n",
      "    }\n",
      "}\n",
      "\n"
     ]
    }
   ],
   "source": [
    "look_at_the_format_of_the_file('C:\\\\Users\\\\jdale\\\\OneDrive\\\\School\\\\Text Analytics\\\\json_schema.txt')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Choose documents"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['4c159330655513984fac5ab8e0c575512d33b514.json',\n",
      " 'eb265b07935811bc052549f3780c9af843ecf45e.json',\n",
      " '8da224b5eba13cee28c3a95cb3c1e3c4d26e5a3a.json',\n",
      " '5e5908b9f7ad23dd332a32e1dfa411c4b0e88f59.json',\n",
      " 'ddb7090314e89cfc001186101093a15896ba3b43.json',\n",
      " 'b142d0d3130f98fe1bc2cec14d62de4362c80534.json',\n",
      " 'd3d1b82a318250a2e7193878c11a4e14817b07eb.json',\n",
      " '49b80f692b4d3e2ebbd59454e10c29d522d8c2f7.json',\n",
      " '9f5625d9287d8c6729215a20add99c8cd4ba08cb.json',\n",
      " '0953fa36903063f60627e07f7b4e07f0aec3c4d3.json']\n"
     ]
    }
   ],
   "source": [
    "sampling_of_files = choose_documents(glob.glob(\n",
    "    'C:\\\\Users\\\\jdale\\\\OneDrive\\\\School\\\\Text Analytics\\\\*\\\\*\\\\pdf_json'), 10)\n",
    "import pprint\n",
    "pprint.pprint(sampling_of_files[:10])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Write a files reader"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['The global assurance of safe drinking water and basic sanitation has been '\n",
      " 'recognised as a United Nations Millennium Development Goal 1 , particularly '\n",
      " 'in light of the pressures of rising urbanisation, agricultural '\n",
      " 'intensification and climate change 2,3 . These trends enforce an increasing '\n",
      " 'demand for freshwater monitoring frameworks that combine cost effectiveness, '\n",
      " 'fast technology deployability and data transparency 4 . Environmental '\n",
      " 'metagenomics, the tracing of organisms present in a substrate through '\n",
      " 'high-throughput DNA sequencing, yields informative measures of relative '\n",
      " 'taxonomic species occurrence and functional diversity 5 . Microbial '\n",
      " 'metagenomics studies overcome enrichment biases common to traditional '\n",
      " 'culturing approaches 5 ; however, they usually depend on expensive and '\n",
      " 'stationary equipment, highly specialised operational training and '\n",
      " 'substantial time lags between fieldwork, sample preparation, raw data '\n",
      " 'generation and access.In recent years, these challenges have been revisited '\n",
      " \"with the prospect of 'portable' DNA analysis. The main driver of this is the \"\n",
      " 'smartphone-sized MinION device from Oxford Nanopore Technologies (ONT), '\n",
      " 'which enables real-time DNA sequencing using nanopores 6 . Nanopore read '\n",
      " 'lengths can be comparably long (currently up to ~2*10 6 bases 7 ), which is '\n",
      " 'enabled by continuous electrical sensing of sequential nucleotides along '\n",
      " 'single DNA strands. In connection with a laptop or cloud access for the '\n",
      " 'translation of raw voltage signal into nucleotides, nanopore sequencing can '\n",
      " 'be used to rapidly monitor long DNA sequences in remote locations. Although '\n",
      " \"there are still common concerns about the technology's base-level accuracy, \"\n",
      " 'mobile MinION setups have already proven powerful for real-time tracing and '\n",
      " 'open data sharing during bacterial and viral pathogen outbreaks [8] [9] [10] '\n",
      " '[11] [12] [13] .Here we report a simple, inexpensive workflow to assess '\n",
      " 'microbial freshwater ecosystems with nanopore DNA sequencing. Our benchmark '\n",
      " 'involves the design and optimisation of essential experimental steps for '\n",
      " 'multiplexed MinION usage in the context of local environments, together with '\n",
      " 'an evaluation of computational methods for the bacterial classification of '\n",
      " 'nanopore sequencing reads from metagenomic libraries. To showcase the '\n",
      " 'resolution of sequencing-based aquatic monitoring in a spatiotemporal '\n",
      " 'setting, we combine DNA analyses with physicochemical measurements of '\n",
      " 'surface water samples collected at nine locations within a confined ~12 '\n",
      " 'kilometre reach of the River Cam passing through the city of Cambridge (UK) '\n",
      " 'in April, June and August 2018.Nanopore full-length (V1-V9) 16S ribosomal '\n",
      " 'RNA (rRNA) gene sequencing was performed on all locationbarcoded freshwater '\n",
      " 'samples at each of the three time points (Figure 1 ; Supplementary Table 1a '\n",
      " '). Samples were complemented with a negative control (deionised water) and a '\n",
      " 'mock community control composed of eight bacterial species in known mixture '\n",
      " 'proportions (Methods).To obtain valid taxonomic assignments from freshwater '\n",
      " 'sequencing profiles using nanopore sequencing, twelve different '\n",
      " 'classification tools were compared through several performance metrics '\n",
      " '(Extended Data Figure 1 ; Methods). Root mean square errors (RMSE) between '\n",
      " 'observed and expected bacteria of the mock community differed slightly '\n",
      " 'across all classifiers. An Enterobacteriaceae overrepresentation was '\n",
      " 'observed across all replicates and classification methods, pointing towards '\n",
      " 'a consistent Escherichia coli amplification bias potentially caused by '\n",
      " 'skewed taxonomic specificities of the selected 16S primer pair (27f and '\n",
      " '1492r) 14 . Robust quantifications were obtained by Minimap2 15 alignments '\n",
      " 'against the SILVA v.132 database 16 , for which 99.68 % of classified reads '\n",
      " 'aligned to the expected mock community taxa (mean sequencing accuracy 92.08 '\n",
      " '%). Minimap2 classifications reached the second lowest RMSE (excluding '\n",
      " 'Enterobacteriaceae), and relative quantifications were highly consistent '\n",
      " 'between mock community replicates. Benchmarking of the classification tools '\n",
      " \"on one aquatic sample further confirmed Minimap2's reliable performance in a \"\n",
      " 'more complex bacterial community, although other tools such as SPINGO 17 , '\n",
      " 'MAPseq 18 , or IDTAXA 19 also produced highly concordant results despite '\n",
      " 'variations in processing speed and memory usage (data not shown).Using '\n",
      " 'Minimap2 classifications within our bioinformatics consensus workflow '\n",
      " '(Extended Data Figure 2 ;Methods), we then inspected sequencing profiles of '\n",
      " 'three independent MinION runs for a total of 30 river DNA isolates and six '\n",
      " 'controls. This yielded ~8. 3 and Rhodobacteraceae (2.5 %) (Figure 2d ). '\n",
      " 'Members of these families are commonly associated with aquatic environments; '\n",
      " 'for example, Burkholderiaceae reads mostly originate from genera such as '\n",
      " 'Limnohabitans, Rhodoferax or Aquabacterium, which validates the suitability '\n",
      " 'of this nanopore metagenomics workflow.Hierarchical clustering additionally '\n",
      " 'showed that two biological replicates collected at the same location and '\n",
      " 'time point (April samples 9.1 and 9.2), grouped with high concordance; this '\n",
      " 'indicates that spatiotemporal trends are discernible even within a highly '\n",
      " 'localised context. Besides the dominant core microbiome, microbial profiles '\n",
      " 'showed a marked arrangement of time dependence, with water samples from '\n",
      " 'April grouping more distantly to those from June and August (Figure 2c ). '\n",
      " 'Principal component analysis (PCA) (Figure 3a ; Extended Data Figure   4 ) '\n",
      " 'revealed that the strongest differential abundances along the chronological '\n",
      " 'axis of variation (PC3) derived from the higher abundance of '\n",
      " 'Carnobacteriaceae in April (Figure 3b ). This family is known for its '\n",
      " 'occurrence in waters with low temperature 20 .While a seasonal difference in '\n",
      " 'bacterial composition can be expected due to increasing water temperatures '\n",
      " 'in the summer months, additional changes may have also been caused by '\n",
      " 'alterations in river hydrochemistry and flow rate (Extended Data Figures 5 '\n",
      " 'and 6 , respectively; Supplementary Table 1c) . To assess this effect in '\n",
      " 'detail, we measured the pH and a range of major and trace cations in all '\n",
      " 'river water samples using inductively coupled plasma-optical emission '\n",
      " 'spectroscopy (ICP-OES), as well as major anions using ion chromatography '\n",
      " '(Extended Data Figure 5 ; Methods). As with the bacterial composition '\n",
      " 'dynamics, we observed significant temporal variation in water chemistry, '\n",
      " 'superimposed on a spatial gradient of generally increasing sodium and '\n",
      " 'chloride concentrations along the river reach. This spatially consistent '\n",
      " 'effect is likely attributed to wastewater and agricultural discharge inputs '\n",
      " 'in and around Cambridge city. A comparison of the major element chemistry in '\n",
      " \"the River Cam transect with the world's 60 largest rivers further \"\n",
      " 'corroborates the likely impact of anthropogenic pollution in this fluvial '\n",
      " 'ecosystem 21 (Extended Data Figure 5 ; Methods).In line with these '\n",
      " 'physicochemical trends, we next determined the spatiotemporal enrichment of '\n",
      " 'potentially functionally important bacterial taxa through nanopore '\n",
      " 'sequencing. We retrieved 55 potentially pathogenic bacterial genera through '\n",
      " 'careful integration of species known to affect human health 22, 23 , and '\n",
      " 'also 13 wastewaterassociated 24 bacterial genera (Supplementary Table 3 ). '\n",
      " 'Of these, 21 potentially pathogenic and eight wastewaterassociated genera '\n",
      " 'were detected across all of the river samples (Figure 3c ; Methods). Many of '\n",
      " 'these signals were stronger downstream of urban sections, within the mooring '\n",
      " 'zone for recreational and residential barges (location 7, Figure 1a ) and in '\n",
      " 'the vicinity of sewage outflow from a nearby wastewater treatment plant '\n",
      " '(location 8). The most prolific candidate pathogen genus observed was '\n",
      " 'Arcobacter, which features multiple species implicated in acute '\n",
      " 'gastrointestinal infections 25 .In general, much of the taxonomic variation '\n",
      " 'across all samples was caused by sample April-7 (PC1 explains 27. Using '\n",
      " 'multiple sequence alignments between nanopore reads and pathogenic species '\n",
      " 'references, we further resolved the phylogenies of three common potentially '\n",
      " 'pathogenic genera occurring in our river samples, Pseudomonas, Legionella '\n",
      " 'and Salmonella (Extended Data Figure 7 ; Methods). While Legionella and '\n",
      " 'Salmonella diversities only presented negligible levels of known harmful '\n",
      " 'species, a cluster of sequencing reads in downstream sections indicated a '\n",
      " 'low abundance of the opportunistic, environmental pathogen Pseudomonas '\n",
      " 'aeruginosa (Extended Data Figure 7 ).We also found significant variations in '\n",
      " 'relative abundances of the Leptospira genus, which was recently described to '\n",
      " 'be enriched in wastewater effluents in Germany 27 . Indeed, the peak of '\n",
      " 'River Cam Leptospira reads falls into an area of increased sewage influx '\n",
      " '(Figure 3c ). The Leptospira genus contains several potentially pathogenic '\n",
      " 'species capable of causing life-threatening leptospirosis through waterborne '\n",
      " 'infections 28 , however also features closerelated saprophytic and '\n",
      " \"'intermediate' taxa 29 . To resolve its complex phylogeny in the River Cam \"\n",
      " 'surface, we aligned Leptospira reads from all samples together with various '\n",
      " 'reference sequences assigned to pre-classified pathogenic, saprophytic and '\n",
      " 'other environmental Leptospira species 29 (Figure 3d ; Supplementary Table '\n",
      " '4a; Methods). Despite the presence of nanopore sequencing errors (Extended '\n",
      " 'Data Figure 8 ) and correspondingly inflated read divergence, we could '\n",
      " 'pinpoint spatial clusters and a distinctly higher similarity between our '\n",
      " 'amplicons and saprophytic rather than pathogenic Leptospira species. These '\n",
      " 'findings were subsequently validated by targeted, Leptospira '\n",
      " 'species-specific qPCR (Supplementary Table 5 , Methods), confirming that the '\n",
      " 'latest nanopore sequencing quality is high enough to yield indicative '\n",
      " 'results for bacterial monitoring workflows at the species level.Using an '\n",
      " 'inexpensive, easily adaptable and scalable framework, we provide the first '\n",
      " 'spatiotemporal nanopore sequencing atlas of bacterial microbiota along a '\n",
      " 'river reach. Beyond the core microbiome of an example fluvial ecosystem, our '\n",
      " 'results suggest that it is possible to robustly assess the heterogeneity in '\n",
      " 'accessory bacterial composition in the context of supporting physical '\n",
      " '(temperature, flow rate) and hydrochemical (pH, inorganic solutes) '\n",
      " \"parameters. We show that the technology's current accuracy of ~92 % allows \"\n",
      " 'for the designation of significant human pathogen community shifts along '\n",
      " 'rural-to-urban river transitions, as illustrated by downstream increases in '\n",
      " 'the abundance of pathogen candidates.Furthermore, our assessment of popular '\n",
      " 'bioinformatics workflows for taxonomic classification highlights current '\n",
      " 'challenges with error-prone nanopore sequences. We observed differences in '\n",
      " 'terms of bacterial quantifications, read misclassification rates and '\n",
      " 'consensus agreements between the twelve tested computational methods. In '\n",
      " 'this computational benchmark, using the SILVA v.132 reference database, one '\n",
      " 'of the most balanced performances was achieved by Minimap2 alignments. As '\n",
      " 'nanopore sequencing quality continues to increase through refined pore '\n",
      " 'chemistries, basecalling algorithms and consensus sequencing workflows [30] '\n",
      " '[31] [32] , future bacterial taxonomic classifications are likely to improve '\n",
      " 'and advance opportunities for aquatic species discovery.We show that '\n",
      " 'nanopore amplicon sequencing data can resolve the core microbiome of a '\n",
      " 'freshwater body, as well as its temporal and spatial fluctuations. Besides '\n",
      " 'common freshwater bacteria, we find that the differential abundances of '\n",
      " 'Carnobacteriaceae most strongly contribute to seasonal loadings in the River '\n",
      " 'Cam.Carnobacteriaceae have been previously associated with cold environments '\n",
      " '20 , and we found these to be more abundant in colder April samples (mean '\n",
      " '11.3 °C, vs. 15.8 °C in June and 19.1 °C in August). This might help to '\n",
      " 'establish this family as an indicator for bacterial community shifts along '\n",
      " 'with water temperature fluctuations.Most routine freshwater surveillance '\n",
      " 'frameworks focus on semi-quantitative diagnostics of only a limited number '\n",
      " 'of target taxa, such as pathogenic Salmonella, Legionella and faecal '\n",
      " 'coliforms 33, 34 . Our proof-of-principle analysis highlights that the '\n",
      " 'combination of full-length 16S rRNA gene amplification and nanopore '\n",
      " 'sequencing can complement hydrochemical controls in pinpointing relatively '\n",
      " 'contaminated freshwater sites, some of which had been previously highlighted '\n",
      " 'for their pathogen diversity and abundance of antimicrobial resistance genes '\n",
      " '35, 36 .Nanopore sequencing here allowed for the reliable distinction of '\n",
      " 'closely related pathogenic and non-pathogenic bacterial species of the '\n",
      " 'common Salmonella, Legionella, Pseudomonas and Leptospira genera. For '\n",
      " 'Leptospira bacteria, we further validated nanopore sequencing results '\n",
      " 'through the gold standard qPCR workflow of Public Health England '\n",
      " '(Supplementary Table 5 ).A number of experimental intricacies should be '\n",
      " 'addressed towards future nanopore freshwater sequencing studies, mostly by '\n",
      " 'scrutinising water DNA extraction yields, PCR biases and molar imbalances in '\n",
      " 'barcode multiplexing ( Figure 2a ; Extended Data Figure 8 ). Yet, our '\n",
      " 'results show that it would be theoretically feasible to obtain meaningful '\n",
      " 'river microbiota from >100 barcoded samples on a single nanopore flow cell, '\n",
      " 'thereby enabling water monitoring projects involving large collections at '\n",
      " 'costs below £20 per sample (Supplementary Table 6 ). Barcoded shotgun '\n",
      " 'nanopore sequencing protocols may pose a viable alternative strategy to '\n",
      " 'bypass pitfalls often observed in amplicon-based workflows, namely '\n",
      " 'taxon-specific primer biases 14 Competing interests: All authors of this '\n",
      " 'manuscript declare no competing interest.Stammnitz (maxrupsta@gmail.com), or '\n",
      " 'to Andre Holzer (andre.holzer.biotech@gmail.com) and Lara Urban '\n",
      " '(lara.h.urban@gmail.com). CC-BY-ND 4.0 International license author/funder. '\n",
      " 'It is made available under a The copyright holder for this preprint (which '\n",
      " 'was not peer-reviewed) is the . https://doi.org/10.1101/2020.02.06.936302 '\n",
      " 'doi: bioRxiv preprint The copyright holder for this preprint (which was not '\n",
      " 'peer-reviewed) is the . https://doi.org/10.1101/2020.02.06.936302 doi: '\n",
      " 'bioRxiv preprint The copyright holder for this preprint (which was not '\n",
      " 'peer-reviewed) is the . https://doi.org/10.1101/2020.02.06.936302 doi: '\n",
      " 'bioRxiv preprint The copyright holder for this preprint (which was not '\n",
      " 'peer-reviewed) is the . https://doi.org/10.1101/2020.02.06.936302 doi: '\n",
      " 'bioRxiv preprint Q1 -1.5*IQR (lower), and Q3 + 1.5*IQR (upper), '\n",
      " 'respectively; Q1: first quartile, Q3: third quartile, IQR: interquartile '\n",
      " 'range. The centre colour-scale table depicts the categorisation of subsets '\n",
      " 'of genera as waterborne bacterial pathogens (WB), drinking water pathogens '\n",
      " '(DWP), potential drinking water pathogens (pDWP), human pathogens (HP) and '\n",
      " 'core genera from wastewater treatment plants (WW) (dark grey: included, '\n",
      " 'light grey: Table 3 ). The right-hand circle plot shows the distribution of '\n",
      " 'bacterial genera across locations of the River Cam. Circle sizes represent '\n",
      " 'overall read size fractions, while circle colours (sigma scheme) represent '\n",
      " 'the standard deviation from the observed mean relative abundance within each '\n",
      " 'genus. (d) Phylogenetic tree illustrating the multiple sequence alignment of '\n",
      " 'all nanopore reads classified as Leptospira, together with known Leptospira '\n",
      " 'reference sequences ranging from pathogenic to saprophytic species 29 '\n",
      " '(Supplementary Table   4a ). The copyright holder for this preprint (which '\n",
      " 'was not peer-reviewed) is the . https://doi.org/10.1101/2020.02.06.936302 '\n",
      " 'doi: bioRxiv preprint MATERIALS AND METHODSWe monitored nine distinct '\n",
      " 'locations along a 11.62 km reach of the River Cam, featuring sites upstream, '\n",
      " 'downstream and within the urban belt of the city of Cambridge, UK. '\n",
      " 'Measurements were taken at three time points, in two-month intervals between '\n",
      " 'April and August 2018 (Figure 1; Supplementary Table 1a ). To warrant river '\n",
      " 'base flow conditions and minimise rain-derived biases, a minimum dry weather '\n",
      " 'time span of 48h was maintained prior to sampling 51 . One litre of surface '\n",
      " 'water was collected in autoclaved DURAN bottles (Thermo Fisher Scientific, '\n",
      " 'Waltham, MA, USA), and cooled to 4 °C within three hours. Two bottles of '\n",
      " 'water were collected consecutively for each time point, serving as '\n",
      " 'biological replicates of location 9 (samples 9.1 and 9.2).We assessed '\n",
      " 'various chemical, geological and physical properties of the River Cam '\n",
      " '(Extended Data Figures 5 and   6 , Supplementary Tables 1b and 1c) .In situ '\n",
      " 'water temperature was measured immediately after sampling. To this end, we '\n",
      " 'linked a DS18B20 digital temperature sensor to a portable custom-built, grid '\n",
      " 'mounted Arduino nano v3. The copyright holder for this preprint (which was '\n",
      " 'not peer-reviewed) is the . https://doi.org/10.1101/2020.02.06.936302 doi: '\n",
      " 'bioRxiv preprint indicated precision of more than 4 % for Cl -. However, the '\n",
      " 'high Clconcentrations of the samples in this study were not fully bracketed '\n",
      " 'by the calibration curve and we therefore assigned a more conservative '\n",
      " 'uncertainty of 10 % to Clconcentrations.High calcium and magnesium '\n",
      " 'concentrations were recorded across all samples, in line with hard '\n",
      " 'groundwater and natural weathering of the Cretaceous limestone bedrock '\n",
      " 'underlying the river catchment (Extended Data Figure 5 ).There are no known '\n",
      " 'evaporite salt deposits in the river catchment, and therefore the high '\n",
      " 'dissolved Na + , K + and Clconcentrations in the River Cam are likely '\n",
      " 'derived from anthropogenic inputs 52 (Extended Data Figure 5 ). We The total '\n",
      " 'dissolved solid (TDS) concentration across the 30 freshwater samples had a '\n",
      " 'mean of 458 mg/L (range 325 -605 mg/L) which is relatively high compared to '\n",
      " 'most rivers, due to 1.) substantial solute load in the Chalk groundwater '\n",
      " '(particularly Ca 2+ , Mg 2+ , and HCO3 -) and 2.) likely anthropogenic '\n",
      " 'contamination (particularly Na + , Cl -, and SO4 2-). The TDS range and the '\n",
      " 'major ion signature of the River Cam is similar to other anthropogenically '\n",
      " 'heavily-impacted rivers 21 , exhibiting enrichment in Na + (Extended Data '\n",
      " 'Figure 5 ).Overall, ion profiles clustered substantially between the three '\n",
      " 'time points, indicating characteristic temporal shifts in water chemistry. '\n",
      " 'PC1 of a PCA on the solute concentrations [µmol/L] shows a strong time '\n",
      " 'effect, separating spring (April) from summer (June, August) samples '\n",
      " '(Extended Data Figure 5b) . We highlighted the ten most important features '\n",
      " '(i.e., features with the largest weights) and their contributions to PC1 '\n",
      " '(Extended Data Figure   5c ).We integrated sensor data sets on mean daily '\n",
      " 'air temperature, sunshine hours and total rainfall from a public, The '\n",
      " 'copyright holder for this preprint (which was not peer-reviewed) is the . '\n",
      " 'https://doi.org/10.1101/2020.02.06.936302 doi: bioRxiv preprintWithin 24 '\n",
      " 'hours of sampling, 400 mL of refrigerated freshwater from each site was '\n",
      " 'filtered through an individual 0.22 µm pore-sized nitrocellulose filter '\n",
      " '(MilliporeSigma, Burlington, MA, USA) placed on a Nalgene polysulfone bottle '\n",
      " 'top filtration holder (Thermo Fisher Scientific) at -30 mbar vacuum '\n",
      " 'pressure. Additionally, 400 mL deionised (DI) water was also filtered. We '\n",
      " 'then performed DNA extractions with a modified DNeasy PowerWater protocol '\n",
      " '(Qiagen, Hilden, Germany). Briefly, filters were cut into small slices with '\n",
      " 'sterile scissors and transferred to 2 mL Eppendorf tubes containing lysis '\n",
      " 'beads. Homogenization buffer PW1 was added, and the tubes subjected to ten '\n",
      " 'minutes of vigorous shaking at 30 Hz in a TissueLyser II machine (Qiagen). '\n",
      " 'After subsequent DNA binding and washing steps in accordance with the '\n",
      " \"manufacturer's protocol, elution was done in 50 µL EB. We used Qubit dsDNA \"\n",
      " 'HS Assay (Thermo Fisher Scientific) to determine water DNA isolate '\n",
      " 'concentrations (Supplementary Table 2a ).DNA extracts from each sampling '\n",
      " 'batch and DI water control were separately amplified with V1-V9 full-length '\n",
      " 'Figure 1b-c) , which was previously assessed using nanopore shotgun '\n",
      " 'metagenomics 42 . We used common primer binding sequences 27f and 1492r, '\n",
      " 'both coupled to unique 24 bp barcodes and a nanopore motor protein tether '\n",
      " 'sequence (Supplementary Table 7 ). Fulllength 16S rDNA PCRs were performed '\n",
      " 'with the following conditions: Table S2b ). We used KAPA Pure Beads (KAPA '\n",
      " 'Biosystems, Wilmington, MA, USA) to concentrate full-length 16S rDNA '\n",
      " 'products in 21 µL DI water. Multiplexed nanopore ligation sequencing '\n",
      " 'libraries were then made by following the SQK-LSK109 protocol (Oxford '\n",
      " 'Nanopore Technologies, Oxford, UK).R9.4 MinION flow cells (Oxford Nanopore '\n",
      " 'Technologies) were loaded with 75 µl of ligation library. The MinION '\n",
      " 'instrument was run for approximately 48 hours, until no further sequencing '\n",
      " 'reads could be collected. Fast5 files were basecalled using Guppy (version '\n",
      " '3.15) and output DNA sequence reads with Q>7 were saved as fastq '\n",
      " 'files.Various output metrics per library and barcode are summarised in '\n",
      " 'Supplementary Table 2c .In collaboration with Public Health England, raw '\n",
      " 'water DNA isolates of the River Cam from each location and time point were '\n",
      " 'subjected to the UK reference service for leptospiral testing (Supplementary '\n",
      " 'Table 5 ). This test is based on quantitative real-time PCR (qPCR) of 16S '\n",
      " 'rDNA and LipL32, implemented as a TaqMan assay for the detection and '\n",
      " 'differentiation of pathogenic and non-pathogenic Leptospira spp. from human '\n",
      " 'serum. Briefly, the assay consists of a two-component PCR; the first '\n",
      " 'component is a duplex assay that targets the gene encoding the outer '\n",
      " 'membrane lipoprotein LipL32, which is reported to be strongly associated '\n",
      " 'with the pathogenic phenotype.The second reaction is a triplex assay '\n",
      " 'targeting a well conserved region within the 16S rRNA gene (rrn) in The '\n",
      " 'described data processing and read classification steps were implemented '\n",
      " 'using the Snakemake workflow management system 53 and are available on '\n",
      " 'Github -together with all necessary downstream analysis scripts to reproduce '\n",
      " 'the results of this manuscript (https://github.com/d-j-k/puntseq). We '\n",
      " 'gathered read statistics such as quality scores and read lengths using '\n",
      " 'NanoStat (version 1.1.2, https://github.com/wdecoster/nanostat), and used '\n",
      " 'Pistis (https://github.com/mbhall88/pistis) to create quality control plots. '\n",
      " 'This allowed us to assess GC content and Phred quality score distributions, '\n",
      " 'which appeared consistent across and within our reads. Overall, we obtained '\n",
      " '2,080,266 reads for April, 737,164 for June, and 5,491,510 for August, with '\n",
      " 'a mean read quality of 10.0 (Supplementary Table 2c ).We used twelve '\n",
      " 'different computational tools for bacterial full-length 16S rDNA sequencing '\n",
      " 'read classification We used nanopore sequencing data from our mock community '\n",
      " 'and freshwater amplicons for benchmarking the classification tools. We '\n",
      " 'therefore subsampled (a) 10,000 reads from each of the three mock community '\n",
      " 'sequencing replicates (section 1.4), and (b) 10,000 reads from an aquatic '\n",
      " 'sample (April-8; three random draws served as replicates). We then used the '\n",
      " 'above twelve classification tools to classify these reads against the same '\n",
      " 'database,SILVA v.132 16 (Extended Data Figure 1 ).For the mock community '\n",
      " 'classification benchmark, we assessed the number of unclassified reads, '\n",
      " 'misclassified reads (i.e. sequences not assigned to any of the seven '\n",
      " 'bacterial families), and the root mean squared error (RMSE) between observed '\n",
      " 'and expected taxon abundance of the seven bacterial families. Following the '\n",
      " 'detection of a strong bias towards the Enterobacteriaceae family across all '\n",
      " 'classification tools, we also analysed RMSE values after exclusion of this '\n",
      " 'family (Extended Data Figure 1b -c).For the aquatic sample, the number of '\n",
      " 'unclassified reads were counted prior to monitoring the performance of each '\n",
      " 'classification tool in comparison with a consensus classification, which we '\n",
      " 'defined as majority vote across classifications from all computational '\n",
      " 'workflows. We observed stable results across all three draws of 10,000 reads '\n",
      " 'from the same dataset (data not shown), indicating a robust representation '\n",
      " 'of the performance of each classifier.Minimap2 performed second best at '\n",
      " 'classifying the mock community (lowest RMSE), while also delivering '\n",
      " 'freshwater bacterial profiles in line with the majority vote of other '\n",
      " 'classification tools (Extended Data Figure 1d e), in addition to providing '\n",
      " 'rapid speed (data not shown). Yet, the application of this software to our '\n",
      " 'entire dataset caused insufficient memory errors (at ~150 Gb RAM with kmer '\n",
      " 'length 12), likely due to major sequence redundancies within the SILVA v.132 '\n",
      " 'reference fasta file. Therefore, to run each of our full samples within a '\n",
      " 'reasonable memory limit of 50 Gb, it was necessary to reduce the number of '\n",
      " \"threads to 1, raise the kmer size ('k') to 15 and set the minibatch size \"\n",
      " \"('-K') to 25M (i.e., the number of query bases that are processed at any \"\n",
      " 'time), prolonging the runtime of several samples to ~three days.After '\n",
      " 'applying Minimap2 to the processed reads as explained above (section 2.2.4), '\n",
      " 'we processed the resulting SAM files by firstly excluding all header rows '\n",
      " \"starting with the '@' sign and then transforming the sets of read IDs, SILVA \"\n",
      " 'IDs, and alignment scores to TSV files of unique read-bacteria assignments '\n",
      " 'either on the bacterial genus or family level. All reads that could not be '\n",
      " 'assigned to the genus or family level were discarded, respectively.In the '\n",
      " 'case of a read assignment to multiple taxa with the same alignment score, we '\n",
      " 'determined the lowest taxonomic level in which these multiple taxa would be '\n",
      " 'included. If this level was above the genus or family level, respectively, '\n",
      " 'we discarded the read.Across three independent sequencing replicates of the '\n",
      " 'same linear bacterial community standard (section 2.2.1),we found that the '\n",
      " 'fraction of reads assigned to unexpected genus level taxa lies at ~1 % when '\n",
      " 'using the Minimap2 classifier and the SILVA v.132 database.Raw quantified '\n",
      " 'DNA, PCR amplicons and sequencing read counts were considerably less '\n",
      " 'abundant in DI water negative controls, as compared to actual freshwater '\n",
      " 'specimens (Supplementary Table 2a Table 8 ).Minimap2 alignments against mock '\n",
      " 'community taxa were used to determine the mean read-wise nanopore sequencing '\n",
      " 'accuracy for this study, as determined by the formula: accuracy = 1 -(read '\n",
      " 'mismatch length ÷ read alignment length)These values were calculated for '\n",
      " 'each of all eight species against each sequencing replicate, using the '\n",
      " 'samtools 66 (v.1.3.1) stats function.Sample-specific rarefaction curves were '\n",
      " 'generated by successive subsampling of sequencing reads classified by Figure '\n",
      " '3 ; section 2.4.1). Although we mainly present a single example rarefied '\n",
      " 'dataset within this manuscript, we repeated each analysis, including PCAs, '\n",
      " 'hierarchical clustering and Mantel tests, based on additional rarefied '\n",
      " 'datasets to assess the stability of all results.We performed Mantel tests '\n",
      " '(using scikit-bio version 0.5.1) to compare rarefied datasets with the full '\n",
      " 'dataset. We All classification assessment steps and summary statistics were '\n",
      " 'performed in R or python (https://github.com/dj-k/puntseq). We used the '\n",
      " \"python package 'scikit-bio' for the calculation of the Simpson index and the \"\n",
      " \"Shannon's diversity as well as equitability index.. CC-BY-ND 4.0 \"\n",
      " 'International license author/funder. It is made available under a The '\n",
      " 'copyright holder for this preprint (which was not peer-reviewed) is the . '\n",
      " 'https://doi.org/10.1101/2020.02.06.936302 doi: bioRxiv preprintRarefied read '\n",
      " 'count data was subjected to a log10(x+1) transformation before hierarchical '\n",
      " 'clustering using the complete linkage method. For PCA analyses, rarefied '\n",
      " 'read count data was subjected to log10(x+1) and Ztransformations. Negative '\n",
      " 'control samples were removed. Mock community samples were initially removed '\n",
      " 'to then be re-aligned to the eigenspace determined by the aquatic samples. '\n",
      " 'We provide PCA visualisations of the main principal components (PCs '\n",
      " 'explaining >10 % variance, respectively). For each of these relevant PCs, we '\n",
      " 'further highlight the ten most important features (i.e., taxa with the '\n",
      " 'largest weights) and their contributions to the PCs in barplots.For '\n",
      " 'detecting outlier bacterial families per sample, we chose bacteria which '\n",
      " 'were 1.) identified by more than 500 reads and 2.) which were at least five '\n",
      " 'times more abundant in any single sample than in the mean of all samples '\n",
      " 'combined.A list of 55 known bacterial pathogenic genera, spanning 37 '\n",
      " 'families, was compiled for targeted sequence testing.This was done through '\n",
      " 'the careful integration of curated databases and online sources, foremost '\n",
      " 'using PATRIC 22 and data on known waterborne pathogens 23 (Supplementary '\n",
      " 'Table 3a ). Additionally, we integrated known genera from a large wastewater '\n",
      " 'reference collection 24 (Supplementary Table 3b ).To identify if DNA reads '\n",
      " 'assigned to Leptospiraceae were more similar to sequence reads of previously '\n",
      " 'identified pathogenic, intermediate or environmental Leptospira species, we '\n",
      " 'built a neighbour-joining tree of Leptospiraceae reads classified in our '\n",
      " 'samples data, together with sequences from reference databases (Figure 3d ; '\n",
      " 'species names and NCBI accession numbers in clockwise rotation around the '\n",
      " 'tree in Supplementary Table 4a ). We matched the orientation of our reads, '\n",
      " 'and then aligned them with 68 Leptospira reference sequences and the '\n",
      " 'Leptonema illini reference sequence (DSM 21528 strain 3055) as outgroup. We '\n",
      " 'then built a neighbour-joining tree using Muscle v.3.8.31 67 (excluding '\n",
      " \"three reads in the 'Other Environmental' clade that had extreme branch \"\n",
      " 'lengths >0.2). The reference sequences were annotated as pathogenic and '\n",
      " 'saprophytic clades P1, P2, S1, S2 as recently described 29 .Additional '\n",
      " 'published river water Leptospira that did not fall within these clades were '\n",
      " \"included as 'Other Environmental' 68 . Similarly, we constructed phylogenies \"\n",
      " 'for the Legionella, Salmonella and Pseudomonas genus, using established '\n",
      " 'full-length 16S reference species sequences from NCBI (Supplementary Table '\n",
      " '4b-d).This study was designed to enable freshwater microbiome monitoring in '\n",
      " 'budget-constrained research environments. Although we had access to basic '\n",
      " 'infrastructure such as pipettes, a PCR and TissueLyser II machine, as well a '\n",
      " 'high-performance laptop, we wish to highlight that the total sequencing '\n",
      " 'consumable costs were held below £4,000 (Supplementary Table 6a ). Here, '\n",
      " 'individual costs ranged at ~£75 per sample (Supplementary Table   6b ). With '\n",
      " 'the current MinION flow cell price of £720, we estimate that per-sample '\n",
      " 'costs could be further reduced to as low as ~£15 when barcoding and pooling '\n",
      " '~£100 samples in the same sequencing run (Supplementary Table   6c ). '\n",
      " 'Assuming near-equimolar amplicon pooling, flow cells with an output of '\n",
      " '~5,000,000 reads can yield well over 37,000 sequences per sample and thereby '\n",
      " 'surpass this conservative threshold applied here for comparative river '\n",
      " 'microbiota analyses.Sequencing datasets generated and analysed during this '\n",
      " 'study are available from the European Nucleotide The are no restrictions on '\n",
      " 'data availability.Our Github repository (https://github.com/d-j-k/puntseq/) '\n",
      " 'provides a Snakemake framework that integrates all data pre-processing '\n",
      " 'steps, and a Singularity that contains all necessary software '\n",
      " '(https://github.com/d-jk/puntseq/tree/master/analysis/). We further provide '\n",
      " 'complete and rarefied SILVA 132 classifications from runs of Minimap2 '\n",
      " '(https://github.com/d-j-k/puntseq/tree/master/minimap2_classifications), '\n",
      " 'which can be directly used as an input for downstream analyses.. CC-BY-ND '\n",
      " '4.0 International license author/funder. It is made available under a The '\n",
      " 'copyright holder for this preprint (which was not peer-reviewed) is the . '\n",
      " 'https://doi.org/10.1101/2020.02.06.936302 doi: bioRxiv preprint (a-b) List '\n",
      " 'of pathogen (a) and wastewater (b) candidate bacterial genera. and August '\n",
      " '(c).. CC-BY-ND 4.0 International license author/funder. It is made available '\n",
      " 'under a The copyright holder for this preprint (which was not peer-reviewed) '\n",
      " 'is the . https://doi.org/10.1101/2020.02.06.936302 doi: bioRxiv preprint',\n",
      " 'Coronavirus (CoV) remains a public health concern 16 years after the '\n",
      " 'outbreak of the severe acute respiratory syndrome coronavirus (SARS-CoV) in '\n",
      " '2003 [1] . The Middle East respiratory syndrome coronavirus (MERS-CoV) '\n",
      " 'emerged in 2012, reemerged in 2015, and is still circulating in the Middle '\n",
      " 'East region, which reminds the international community that the threat of '\n",
      " 'CoVs persists [2, 3] . However, neither vaccine nor drugs against CoVs are '\n",
      " 'currently available. Outbreaks of CoVs initiated extensive structural '\n",
      " 'investigation on CoV encoded proteins thereafter, which not only shed light '\n",
      " 'on the life cycle of CoVs but also laid foundation for the structure-based '\n",
      " 'drug design (SBDD). CoV contains a positive single-stranded RNA genome of~30 '\n",
      " 'kb, one of the largest among +RNA viruses [4, 5] . To maintain the unusually '\n",
      " 'large RNA genome, CoV encodes two replicase polyproteins pp1a and pp1ab, '\n",
      " 'which are broken down into 16 nonstructural proteins (nsps) via proteinase '\n",
      " 'cleavage [6, 7] . The nsps are then recruited to cytoplasm membranes, on '\n",
      " 'which they form the membraneassociated replication-transcription complex '\n",
      " '(RTC). An RNAdependent RNA polymerase nsp12 and a helicase nsp13 are the '\n",
      " 'central components of RTC [8, 9] . However, while high-resolution structures '\n",
      " 'of most CoV encoded proteins had been determined soon after SARS-CoV '\n",
      " 'outbreak, the first CoV nsp13 structure, MERS-CoV nsp13, was only solved '\n",
      " 'recently [10] . Nsp13 belongs to helicase superfamily 1 and shares conserved '\n",
      " 'features with the eukaryotic Upf1 helicase [11, 12] . Nsp13 is a '\n",
      " 'multi-domain protein comprising of an N-terminal Cys/His rich domain (CH '\n",
      " 'domain) and a C-terminal SF1 helicase core [10] . Nsp13 exhibits multiple '\n",
      " 'enzymatic activities, including hydrolysis of NTPs and dNTPs, unwinding of '\n",
      " 'DNA and RNA duplexes with 5 0 -3 0 directionality and the RNA 5 0 '\n",
      " '-triphosphatase activity [13, 14] . To investigate the structure of CoV '\n",
      " 'nsp13, we overexpressed the full-length MERS-CoV nsp13 (1-598aa) in insect '\n",
      " 'cells and purified. The activity of the recombinant MERS-CoV nsp13 was '\n",
      " 'verified by ATPase and helicase assays. Crystallization of MERS-CoV nsp13 '\n",
      " 'was achieved by adding a synthetic single-stranded 15 poly dT DNA with 5 0 '\n",
      " '-triphosphate (ppp-15 T) to the protein, which restrains the intrinsic '\n",
      " 'flexibility of nsp13. Benefiting from the presence of an N-terminal '\n",
      " 'zinc-binding domain with three zinc atoms, multi-wavelength anomalous '\n",
      " 'diffraction (MAD) data at the zinc absorption edge was collected, which '\n",
      " 'allowed the determination of the crystal structure of MERS-CoV nsp13 [10] '\n",
      " '.Prepare all solutions using ultrapure water (prepared by purifying '\n",
      " 'deionized water, to attain a sensitivity of 18 MΩ-cm at 25 C) and analytical '\n",
      " 'grade reagents. Prepare and store all reagents at room temperature (unless '\n",
      " 'indicated otherwise). Diligently follow all waste disposal regulations when '\n",
      " 'disposing waste materials. We do not add sodium azide to reagents. 2. The '\n",
      " 'forward primer (gaaattggatccgctgtcggttcatgc) and the reverse primer '\n",
      " '(gaaattctcgagtcactggagcttgtaatt) of full-length nsp13. Primers stocks are '\n",
      " 'either supplied or diluted by molecular biology grade water to 100 μM and '\n",
      " 'stored at À20 C.3. The pFastbac-1 baculovirus transfer vector is modified; 6 '\n",
      " 'Â Histidine-SUMO tag with a C terminal PreScission protease (PPase) site '\n",
      " 'coding sequence in the N terminal of open reading frame [15] .All procedures '\n",
      " 'should be carried out at room temperature unless otherwise '\n",
      " 'specified.Transposition in E. coli DH10 Bac 1. Amplify MERS-nsp13 '\n",
      " 'full-length by PCR method with BamHI and XhoI restriction sites at 5 0 and 3 '\n",
      " '0 termini, respectively.2. The amplified MERS-nsp13 gene should be digested '\n",
      " 'by BamHI/XhoI at 37 C for 1 h. The pFast-bac-6ÂHistidine-SUMO plasmid should '\n",
      " 'also be digested by BamHI/XhoI at the same time.3. Digested nsp13 DNA should '\n",
      " 'be ligated with pFast-bac-6ÂHistidine-SUMO vector using the rapid DNA '\n",
      " 'ligation kit. 1. Prepare 50 mL high-5 cells in express-5 medium at a density '\n",
      " 'of 0.38 Â 10 6 cells/mL, and culture in a 300 mL cell conical flask. '\n",
      " 'Incubate the culture at 28 C with shaking at 120 rpm for 48 h, the density '\n",
      " 'of cells will grow to 1.5-2.5 Â 10 6 cells/mL (see Note 2).2. Add 1.5 mL '\n",
      " 'MERS-CoV nsp13 P2 or P3 virus into the culture, and incubate at 22 C with '\n",
      " 'shaking at 120 rpm for 44-60 h.3. Centrifuge the culture at 3000 Â g for 30 '\n",
      " 'min. Collect the cells pellet. 10. Prepare the Ni-NTA resin, and add the '\n",
      " 'resin into 2-3 empty Econo-Columns (5 mL 50% resin per column), wash and '\n",
      " 'balance the resin with 10 mL lysis and wash buffer(II) twice.11. Place the '\n",
      " 'columns at 4 C. Apply the clarified cell lysates supernatant to the balanced '\n",
      " 'Ni-NTA resin, and flow through the column by gravity.12. Wash the resin in '\n",
      " 'the column with 10 mL lysis and wash buffer (II) 3 times.13. Resuspend the '\n",
      " 'resin by 3.5 mL L lysis and wash buffer(II), and add 100 μL PPase. Incubate '\n",
      " 'the resin at 4 C for 10-12 h.14. Apply the buffer to the column and let it '\n",
      " 'flow under gravity. Collect the flow through in a 50 mL tube.15. Add another '\n",
      " '25 mL lysis and wash buffer(II) to the resin and flow through the column. '\n",
      " 'Also collect the flow through in the previous 50 mL tube. 16 . To remove the '\n",
      " 'PPase, add the flow through to another column which contains the NS4B resin. '\n",
      " 'Collect the flow through from the NS4B resin column.17. Apply the flow '\n",
      " 'through to an Amicon Ultra protein concentrator (30 kDa filter, 50 mL), '\n",
      " 'centrifuge at 2465 Â g at 4 C until the sample volume is concentrated to 1 '\n",
      " 'mL. 18 . Transfer the concentrated sample to a 1.5 mL tube and centrifuge at '\n",
      " '17,949 Â g for 3 min to remove the aggregates and particulates.19. Load the '\n",
      " 'sample onto the superdex 200 column in the size exclude chromatography (SEC) '\n",
      " 'buffer using an Ä KTA-purify chromatography at 4 C.20. Analyze 8 μL of each '\n",
      " 'peak fractions by SDS-PAGE (Fig. 2) . 21 . Collect the fractions that '\n",
      " 'contain the single band of MERS-CoV nsp13, mix the fractions, and '\n",
      " 'concentrate the mixture to a final density of 6-8 mg/mL. 22 . 50 μL packaged '\n",
      " 'the protein sample, quickly freeze them by liquid nitrogen and store them at '\n",
      " 'À80 C. 5. Spot 1 μL sample from the mixture on the thin-layer chromatography '\n",
      " 'cellulose TLC plates and resolve with running buffer for 20 min.6. Dry the '\n",
      " 'plates and press the plate onto phosphor screen for 2 h. Analyze the result '\n",
      " 'by storage phosphor screen and Typhoon Trio Variable Mode Imager (Fig. 3) . '\n",
      " '5. Take 4 μL sample from each reaction mixtures and load the samples onto '\n",
      " '10% native PAGE gel.6. Run the native PAGE gel at 100 V for 40 min on ice.7. '\n",
      " 'Scan the gel (Fig. 4 ).Crystals of the unliganded MERS-CoV nsp13 diffracted '\n",
      " 'the X-rays poorly, >3.6 Å . The addition of 5 0 -triphosphate-15 dT DNA '\n",
      " '(ppp-15T) greatly improves the resolution. 2. Mix 1 μL sample with 1 μL '\n",
      " 'reservoir buffer from the crystallization conditions screen kits, and '\n",
      " 'incubate at 18 C using the hanging-drop vapor-diffusion system.3. '\n",
      " 'Crystallize MERS-CoV nsp13 by mixing with the equal volume of reservoir '\n",
      " 'buffer containing 0.1 M Tris-HCl (pH 8.5), 1 M (NH4) 2 SO 4 , and 15% '\n",
      " 'glycerol. Crystals grow to their maximum in a week (Fig. 5 ).Crystal '\n",
      " 'Structure 1. Highly redundant multi-wavelength anomalous diffraction data '\n",
      " 'should be collected using the X-ray with wavelengths close to the absorption '\n",
      " 'edge of zinc. High energy remote wavelength should be 1.2810 Å , peak '\n",
      " 'wavelength: 1.2827 Å (two datasets were collected to improve the '\n",
      " 'redundancy), and inflection wavelength 1.2831 Å .2. Data processing and '\n",
      " 'reducing by XDS Package and Truncate software from CCP4. The crystals belong '\n",
      " 'to the space group P6 1 22, and contained two copies of nsp13 per asymmetric '\n",
      " 'unit. 3. An interpretable electron density map should be calculated using '\n",
      " 'SHARP/autoSHARP [20] .Coot [21] .5. Collect native data with highest '\n",
      " 'resolution (3.0 Å ) using the X-rays with the wavelength of 0.978 Å .6. '\n",
      " 'Higher resolution structure should be solved by molecular replacement using '\n",
      " 'the initial nsp13 structure as the searching model. 7. Manual model building '\n",
      " 'with the improved electron density map. While most part of nsp13 can be '\n",
      " 'located, the electron density of 1B subdomain is very weak, reflecting that '\n",
      " 'this part is highly flexible.8. Structure refinement to resolution limit of '\n",
      " '3.0 Å using software PHENIX [22] .In the final model (Fig. 6) , 145-230aa '\n",
      " '(the entire 1B domain) of molecule A are disordered, probably due to '\n",
      " 'mobility of 1B and the lack of crystal contacts, whereas in molecule B, 591 '\n",
      " 'out of 598 amino acids were located in the electron density maps (Fig. 7) . '\n",
      " 'Data collection and refinement statistics are summarized in Table 1 . 4 '\n",
      " 'Notes 1. When we prepare P1 virus in six-well plates, the medium in the '\n",
      " 'wells always evaporated. Sealing the gap of the plate by medical tape can '\n",
      " \"reduce the evaporation of medium (don't seal the gap completely, leave a \"\n",
      " 'small gap to keep the ventilation). Having a water trough in incubator also '\n",
      " 'can reduce the evaporation of the medium.2. The culture of insect cells '\n",
      " 'sometimes was harassed by the contamination of bacteria or other microbes. '\n",
      " 'To avoid the contamination, we treat the conical flasks not only by '\n",
      " 'conventional autoclave sterilization, but also leave the 3 L conical flask '\n",
      " '(sealed by tinfoil) in the oven at 200 C for 3-5 h before using.3. To remove '\n",
      " 'nucleic acids bound to nsp13, we used the lysis buffer containing high '\n",
      " 'concentrate salt; this is a key step and improves the crystallization of '\n",
      " 'nsp13 [10] . In practice, when sonicated in the buffer containing high '\n",
      " 'concentrate salt, we found that the SUMO-tagged recombinant proteins lead '\n",
      " 'the supernatant of the high-5 cell lysate to be turbid, which finally blocks '\n",
      " 'the affinity columns. We have tried four concentrations of NaCl in lysis '\n",
      " 'buffer, including 300 mM, 500 mM, 1 M, and 1.5 M. The first three '\n",
      " \"concentrations of NaCl render the supernatant to be unable to use, we can't \"\n",
      " \"improve it by highspeed centrifugation (47,850 Â g), and it also can't be \"\n",
      " 'filtered by 0.45 μm syringe filter. The last concentration, 1.5 M NaCl in '\n",
      " 'lysis buffer, could generate a bit better supernatant of cell lysates than '\n",
      " 'other three concentrations of salt. We centrifuge the supernatant twice, '\n",
      " 'then can filter it by 0.45 μm syringe filters (100 mL supernatant consumed '\n",
      " 'about 8-10 filters). This clarified supernatant can flow through the '\n",
      " 'affinity columns well.4. The results of helicase assay always face the '\n",
      " 'contamination of background fluorescence. Keep the gel from contacting any '\n",
      " 'items containing fluorescence in the lab, including fluorescent dyes, some '\n",
      " 'plastic boxes, hand towel, and so on.']\n"
     ]
    }
   ],
   "source": [
    "string_of_files_list = files_reader(sampling_of_files)\n",
    "pprint.pprint(string_of_files_list[:2])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Cluster documents"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of clusters: 2, Silhouette Score: 0.9270759361330172\n",
      "Number of clusters: 3, Silhouette Score: 0.9207116157180787\n",
      "Number of clusters: 4, Silhouette Score: 0.4672383758046355\n",
      "Number of clusters: 5, Silhouette Score: 0.45250612718556243\n",
      "Number of clusters: 6, Silhouette Score: 0.4552221519281297\n",
      "Number of clusters: 7, Silhouette Score: 0.35165373560844365\n"
     ]
    }
   ],
   "source": [
    "cv, tokenized_files = cluster_documents(string_of_files_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CLUSTER #1\n",
      "['the', 'of', 'and', 'in', 'to', 'is', 'with', 'di', 'or', 'for']\n",
      "CLUSTER #2\n",
      "['the', 'of', 'and', 'in', 'to', 'with', 'for', 'is', 'that', 'was']\n"
     ]
    }
   ],
   "source": [
    "best_cluster(cv, tokenized_files, 2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Summarize documents\n",
    "## Write summarys to file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "summarize_documents(string_of_files_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This file was generated using TextRank summarization. The process taken was to extract a random percent sampling of pdf_json files from https://www.kaggle.com/allen-institute-for-ai/CORD-19-research-challenge. With these files extracted, I then opened all files, tokenized the sentences of the files using tfid vectorizer, and then applied a TextRank algorithm to the tokenized docs. This TextRank allowed me to then extract the 8 most useful summarized sentences. And voila!\n",
      "\n",
      "\n",
      "\n",
      "['The protein expressed from the IBV S1 protein multi-epitope cassette and the NDV NP protein were found at 35 and 53 kDa bands in P1 and P20 of rNDV-IBV-T/B lysates, however, the parental virus strain LaSota only had the~53 kDa NP protein band (Figure 2) , confirming that rNDV-IBV-T/B successfully expressed the IBV S1 protein multi-epitope cassette in vitro and that the protein was antigenic, as evidenced by its recognition on a Western blot by anti-IBV serum.'\n",
      "\n",
      " 'To determine whether inserting the IBV S1 protein multi-epitope cassette into the NDV LaSota genome affects the biological properties of the recombinant virus, the growth characteristics and pathogenicity after 20 passages of rNDV-IBV-T/B were evaluated through virus titration, MDT, and ICPI tests in vitro and in vivo.'\n",
      "\n",
      " 'To determine whether inserting the IBV S1 protein multi-epitope cassette into the NDV LaSota genome affects the biological properties of the recombinant virus, the growth characteristics and pathogenicity after 20 passages of rNDV-IBV-T/B were evaluated through virus titration, MDT, and ICPI tests in vitro and in vivo.'\n",
      "\n",
      " 'The HA titers of LaSota and rNDV-IBV-T/B were the same (2 10 ).'\n",
      "\n",
      " 'The VN antibodies induced by rNDV-IBV-T/B administration could effectively neutralize heterologous IBV strain, indicating that the recombinant vaccine could protect against heterologous IBV challenge.To evaluate the efficacy of the multi-epitope vaccine in protecting against IBV and NDV, one-day-old SPF chicks were immunized with rNDV-IBV-T/B virus or PBS via the oculonasal route.'\n",
      "\n",
      " 'The protection of chickens vaccinated with the multi-epitope vaccine was significantly higher than that of the PBS and LaSota vaccine groups (p < 0.05).'\n",
      "\n",
      " 'The hemagglutination inhibition (HI) titer against NDV and virus neutralization (VN) antibody titer against IBV are expressed as mean log2 Â± standard deviation.To evaluate the efficacy of the multi-epitope vaccine in protecting against IBV and NDV, oneday-old SPF chicks were immunized with rNDV-IBV-T/B virus or PBS via the oculonasal route.'\n",
      "\n",
      " 'The protection of chickens vaccinated with the multi-epitope vaccine was significantly higher than that of the PBS and LaSota vaccine groups (p < 0.05).']\n",
      "\n",
      "\n",
      "\n",
      "['Only mannosylated E glycoproteins (which are exposed at the surface of DV virions transmitted to humans by infected mosquitoes), and not E proteins with complex glycosylation (produced in mammalian cells), have been shown to interact with DC-SIGN-expressing cells (22) .DC-SIGN and L-SIGN are endocytic receptors and their cytoplasmic tails carry putative internalisation signals such as a dileucine (LL) motif (which is present in both DC-SIGN and L-SIGN) and a tri-acidic cluster that is believed to be involved in intracellular trafficking (Fig.'\n",
      "\n",
      " 'We recommend the two following reviews for an overview of the physiological importance of LSECs in viral infections of the liver (6,7) .In this chapter, we will provide general protocols to study the molecular interactions between viruses and DC-SIGN or L-SIGN and to investigate the functions of these two lectins in viral infection and transmission.'\n",
      "\n",
      " 'We next present protocols to produce soluble viral envelope proteins using a Semliki forest virus (SFV) vector (35-37) and to study the molecular basis of viral capture and internalization by DC-SIGN or L-SIGN.'\n",
      "\n",
      " 'Note that only HIV gp120 DMJ , which carries only mannosylated N-glycans, binds to DC-SIGN.Wild-type viruses or viral particles carrying the reporter genes firefly luciferase (Luc) or green fluorescent protein (GFP) can both be used to study DC-SIGN and L-SIGN-mediated cis-infection and trans-enhancement of target cell infection.'\n",
      "\n",
      " 'DC-SIGN-or L-SIGN-expressing cells and their parental counterpart (10 5 cells) are exposed to viral particles for 2 h at 37Â°C at varying MOI in FSC-free medium supplemented with 1% penicillin/streptomycin, pH approx 7.5.'\n",
      "\n",
      " 'Viral transmission is quantified two or 3 d later.DC-SIGN-or L-SIGN-expressing cells and their parental counterpart (10 5 cells) are incubated with viral particles at a low MOI (insufficient to directly infect target cells) for 2 h at 37Â°C and immediately co-cultured with target cells without washing.'\n",
      "\n",
      " 'The protocol is similar to that described under Subheading 3.4.2. or 3.4.3. except that co-culture with target cells is started several days after exposure of cells expressing DC-SIGN or L-SIGN to virus.'\n",
      "\n",
      " 'The advantage of production of viral particles is that cells can be infected at equal MOI and hence the quantity of protein produced is more reproducible than direct electroporation with pSFV-âˆ†env RNA.']\n",
      "\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "file = open('C:\\\\Users\\\\jdale\\\\OneDrive\\\\School\\\\Text Analytics\\\\project_2\\\\SUMMARY.md', 'r')\n",
    "with open(\"C:\\\\Users\\\\jdale\\\\OneDrive\\\\School\\\\Text Analytics\\\\project_2\\\\SUMMARY.md\") as f:\n",
    "    head = [next(f) for i in range(20)]\n",
    "for i in head:\n",
    "    print(i)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
